### ISSUE TITLE
matches_re wrong result

### ISSUE THREAD

**Issue by lixiaolx** [rapidsai/cudf#7836]
**Describe the bug**
When using Matches_re to match the regular expression at the end of $, an incorrect result appears.

**Steps/Code to reproduce bug**
When testing $ (character terminator), I encountered a bug that cudf returns an incorrect result. Test test case:
Test case file:   cudf/cpp/tests/strings/contains_tests.cpp
Test code:
![image](https://user-images.githubusercontent.com/17673134/113382075-30145500-93b3-11eb-986b-89d1240b1ba0.png)
result:
![image](https://user-images.githubusercontent.com/17673134/113382104-40c4cb00-93b3-11eb-8f6c-f2e15ba06247.png)

**some others**
1. What are the differences between the current cudf and java language when constructing regular expressions? If there are differences, can you help me list it?

2. I want to participate in the perfect function of cudf regular expressions, but I find that this piece of implementation is too hard, and there are many implementation details through the code that I don’t understand. In addition to this document（https://github.com/rapidsai/cudf/blob/branch-0.19/cpp/doxygen/regex.md）, can you provide further detailed documents, for example (can you currently developer of provides the detailed implementation logic of the current regularity; what functions are not supported, etc.)

**Comment by kkraus14** [comment]
@lixiaolx based on your code above I don't think your expected results are correct. The expected regex pattern you have here in using `matches_re` is from the beginning of the string:
- Look if the first character is numeric and 1 through 9
- Look if the second character is numeric and 0 through 9
- Look if it's the end of the string

In the [2] element of "123", the first condition would be true, the second condition would be true, but the last condition would be false since it's not the end of the string. In the [4] element of "AB22", the first condition would be false since `matches_re` is from the beginning of the string.

The [7] element of "10\r\n" is a bit interesting to me.

@davidwendt it looks like the `\n` characters here are ignored / allowed to be at the end of the string and it doesn't cause it to return False based on the `$` terminator, but it doesn't work the same way with `\r\n`. What's the expected behavior here?

**Comment by lixiaolx** [comment]
> The expected regex pattern you have here in using [`matches_re`](https://docs.rapids.ai/api/libcudf/stable/group__strings__contains.html#gaef9058cb47b4173e675fd38dd86daa20) is from the beginning of the string:

Yeah, thank you. According to your suggestion, I have changed mathes into contains and got different results, but the last result still didn't match the correct result
test code：
![image](https://user-images.githubusercontent.com/17673134/113389819-3f9b9a00-93c3-11eb-9a43-6294c48feb55.png)

result：
![image](https://user-images.githubusercontent.com/17673134/113389840-475b3e80-93c3-11eb-8f49-339643528834.png)

**Comment by kkraus14** [comment]
I'm guessing it has to do with the Windows line endings of `\r\n`, but I'm not sure. Hopefully someone else can chime in with more insights.

**Comment by davidwendt** [comment]
The CRLF (`\r\n`) is not expected as an end-of-line delimiter in the libcudf regex code logic. Actually, the `\r` itself has no special meaning from what I can see. The behavior also appears to match the Python `re` library behavior:

```
>>> import re
>>> re.match('[1-9][0-9]$','12')
<re.Match object; span=(0, 2), match='12'>

>>> print('match' if re.match('[1-9][0-9]$','11\n') else 'no match')
match
>>> print('match' if re.match('[1-9][0-9]$','11\r') else 'no match')
no match
>>> print('match' if re.match('[1-9][0-9]$','11\r\n') else 'no match')
no match

```
I got the same result using `re.MULTILINE` flag and `re.search` as well.
The patterns above also do not match for `std::regex` and `boost::regex` and `java.util.regex` though to be fair, these did not match on the `11\n` pattern either.

I'm not a regex expert but googling this I did found some suggesting to include the `\r\n` in the pattern somehow instead.
Example: `[1-9][0-9]\r?\n?$`.
I'm not sure if that helps in your use case or not. 
Also it looks like doing a replace of `\r\n` with `\n` on the the column before using the `contains_re/match_re` may work as well.

The regex implementation inside libcudf is based on the [plan9 source code](https://9fans.github.io/plan9port/unix/) in case that helps. 

I don't see any easy way to add support for the CRLF pattern to the libcudf regex code.
